plugins/redfish/fu-redfish-smbios.c

```diff
@@ -0,0 +1,461 @@
+{
+	g_free (self->ip_addr);
+	self->ip_addr = g_strdup (ip_addr);
+}
+
+static void
+fu_redfish_smbios_set_port (FuRedfishSmbios *self, guint16 port)
+{
+	self->port = port;
+}
+
+static void
+fu_redfish_smbios_export (FuFirmware *firmware,
+			FuFirmwareExportFlags flags,
+			XbBuilderNode *bn)
+{
+	FuRedfishSmbios *self = FU_REDFISH_SMBIOS (firmware);
+	fu_xmlb_builder_insert_kx (bn, "port", self->port);
+	fu_xmlb_builder_insert_kv (bn, "hostname", self->hostname);
+	fu_xmlb_builder_insert_kv (bn, "mac_addr", self->mac_addr);
+	fu_xmlb_builder_insert_kv (bn, "ip_addr", self->ip_addr);
+	fu_xmlb_builder_insert_kx (bn, "vid", self->vid);
+	fu_xmlb_builder_insert_kx (bn, "pid", self->pid);
+}
+
+static gboolean
+fu_redfish_smbios_build (FuFirmware *firmware, XbNode *n, GError **error)
+{
+	FuRedfishSmbios *self = FU_REDFISH_SMBIOS (firmware);
+	const gchar *tmp;
+	guint64 tmpu;
+
+	/* optional properties */
+	tmpu = xb_node_query_text_as_uint (n, "port", NULL);
+	if (tmpu != G_MAXUINT64)
+		fu_redfish_smbios_set_port (self, (guint16) tmpu);
+	tmpu = xb_node_query_text_as_uint (n, "vid", NULL);
+	if (tmpu != G_MAXUINT64)
+		self->vid = (guint16) tmpu;
+	tmpu = xb_node_query_text_as_uint (n, "pid", NULL);
+	if (tmpu != G_MAXUINT64)
+		self->pid = (guint16) tmpu;
+	tmp = xb_node_query_text (n, "hostname", NULL);
+	if (tmp != NULL)
+		fu_redfish_smbios_set_hostname (self, tmp);
+	tmp = xb_node_query_text (n, "mac_addr", NULL);
+	if (tmp != NULL)
+		fu_redfish_smbios_set_mac_addr (self, tmp);
+	tmp = xb_node_query_text (n, "ip_addr", NULL);
+	if (tmp != NULL)
+		fu_redfish_smbios_set_ip_addr (self, tmp);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_redfish_smbios_parse_interface_data (FuRedfishSmbios *self,
+					GBytes *fw, gsize offset,
+					GError **error)
+{
+	gsize bufsz = 0;
+	guint8 interface_type = 0x0;
+	guint8 mac_addr[6] = { 0x0 };
+	const guint8 *buf = g_bytes_get_data (fw, &bufsz);
+	g_autofree gchar *mac_addr_str = NULL;
+
+	/* parse the data depending on the interface type */
+	if (!fu_common_read_uint8_safe (buf, bufsz, offset,
+					&interface_type, error))
+		return FALSE;
+	offset++;
+	switch (interface_type) {
+	case REDFISH_INTERFACE_TYPE_USB_NETWORK:
+		if (!fu_common_read_uint16_safe (buf, bufsz, offset + 0x00,
+						 &self->vid, G_LITTLE_ENDIAN, error))
+			return FALSE;
+		if (!fu_common_read_uint16_safe (buf, bufsz, offset + 0x02,
+						 &self->pid, G_LITTLE_ENDIAN, error))
+			return FALSE;
+		g_debug ("USB Network Interface: 0x%04x:0x%04x", self->vid, self->pid);
+		break;
+	case REDFISH_INTERFACE_TYPE_PCI_NETWORK:
+		if (!fu_common_read_uint16_safe (buf, bufsz, offset + 0x00,
+						 &self->vid, G_LITTLE_ENDIAN, error))
+			return FALSE;
+		if (!fu_common_read_uint16_safe (buf, bufsz, offset + 0x02,
+						 &self->pid, G_LITTLE_ENDIAN, error))
+			return FALSE;
+		g_debug ("PCI Network Interface: 0x%04x:0x%04x", self->vid, self->pid);
+		break;
```

Comment:
this seems pretty much the same, can you simplify to one case and have g_debug ("%s Network interface:...", `interface_type == foo ? bar : baz`?

Reply:
The offsets are quite different in v1 and v2 unfortunately :(

Reply:
right but PCI and USB in v1 look the same.

Reply:
Done.